### PR TITLE
DeepGraphInfomax GNNModular pretraining task

### PR DESCRIPTION
**Pull request by tonydavis629**
## Description

Adds in the unsupervised deepgraphinfomax pretraining task for GNNModular. This maximizes the mutual information between positive (same graph) and negative (different graphs) graph representations.

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/models/torch_models/gnn.py [hunk 1]
```diff
@@ -178,6 +180,36 @@ def forward(self, data):
         return out
 
 
+class Discriminator(nn.Module):
+    """
+    This discriminator module is a linear layer without bias, used to measure the similarity between local node representations (`x`) and global graph representations (`summary`).
+
+    The goal of the discriminator is to distinguish between positive and negative pairs of local and global representations.
```
Comment (rbharath) [inline deepchem/models/torch_models/gnn.py:187]: Can you add a usage example here?

File: deepchem/models/torch_models/gnn.py [hunk 2]
```diff
@@ -178,6 +180,36 @@ def forward(self, data):
...
+    """
+    def __init__(self, hidden_dim):
```
Comment (rbharath) [inline deepchem/models/torch_models/gnn.py:189]: Can you document the parameter here in the numpydoc style?

File: deepchem/models/torch_models/gnn.py [hunk 3]
```diff
@@ -178,6 +180,36 @@ def forward(self, data):
...
+
+    def reset_parameters(self):
+        size = self.weight.size(0)
+        uniform(size, self.weight)
+
+    def forward(self, x, summary):
+        """
+        The forward method takes two inputs, `x` (local node representations) and `summary` (global graph representations), both of shape `(batch_size, hidden_dim)`.
+        It computes the product of `summary` and `self.weight`, and then calculates the element-wise product of `x` and the resulting matrix `h`.
+        Finally, it returns the sum of the element-wise product along dimension 1 (i.e., summing over the `hidden_dim`), resulting in a tensor of shape `(batch_size,)`, which represents the similarity scores between the local and global representations.
```
Comment (rbharath) [inline deepchem/models/torch_models/gnn.py:207]: Can you add a numpydoc style Parameters field here?

File: deepchem/models/torch_models/gnn.py [hunk 4]
```diff
@@ -462,6 +507,25 @@ def masked_edge_loss(self, inputs):
 
         return self.edge_mask_loss(pred_edge, inputs)
 
+    def infomax_loss(self, inputs):
+        """
+        Loss that maximizes mutual information between local node representations and a pooled global graph representation. The positive and negative scores represent the similarity between local node representations and global graph representations of simlar and dissimilar graphs, respectively.
```
Comment (rbharath) [inline deepchem/models/torch_models/gnn.py:512]: Can you add a parameters field here?

File: deepchem/models/torch_models/gnn.py [hunk 5]
```diff
@@ -759,3 +823,20 @@ def mask_edges(input_graph: BatchGraphData,
...
+    """
+    Creates a 1-dimensional tensor of integers with a specified length (`num`) and a cyclic shift (`shift`). The tensor starts with integers from `shift` to `num - 1`, and then wraps around to include integers from `0` to `shift - 1` at the end.
+
+    Example
+    -------
+    >>> num = 10
+    >>> shift = 3
+    >>> arr = cycle_index(num, shift)
+    >>> print(arr)
+    tensor([3, 4, 5, 6, 7, 8, 9, 0, 1, 2])
```
Comment (rbharath) [inline deepchem/models/torch_models/gnn.py:838]: Can you add a parameters field here?

File: deepchem/models/losses.py [hunk 6]
```diff
@@ -1112,6 +1112,58 @@ def loss(pred_edge, inputs):
         return loss
 
 
+class DeepGraphInfomaxLoss(Loss):
```
Comment (rbharath) [inline deepchem/models/losses.py:1115]: Can you add a dedicated unit test for this loss?

File: deepchem/models/torch_models/gnn.py [hunk 7]
```diff
@@ -178,6 +180,36 @@ def forward(self, data):
         return out
 
 
+class Discriminator(nn.Module):
```
Comment (rbharath) [inline deepchem/models/torch_models/gnn.py:183]: Can you add a unit test for this layer?

File: deepchem/models/torch_models/gnn.py [hunk 8]
```diff
@@ -178,6 +180,36 @@ def forward(self, data):
...
+    """
```
Comment (rbharath) [inline deepchem/models/torch_models/gnn.py:184]: Can you add this layer to the list of layers in `layers.rst`? We may want to rename this to be something more specific like `LocalGlobalDiscriminator`

**Review by rbharath — COMMENTED** [review]
Some comments mostly around modification to the docs and requests for more unit tests

File: deepchem/models/torch_models/gnn.py (hunk 1 above)
Comment (tonydavis629) [reply deepchem/models/torch_models/gnn.py:187]: done

File: deepchem/models/torch_models/gnn.py (hunk 2 above)
Comment (tonydavis629) [reply deepchem/models/torch_models/gnn.py:189]: done

File: deepchem/models/torch_models/gnn.py (hunk 3 above)
Comment (tonydavis629) [reply deepchem/models/torch_models/gnn.py:207]: done

File: deepchem/models/torch_models/gnn.py (hunk 4 above)
Comment (tonydavis629) [reply deepchem/models/torch_models/gnn.py:512]: done

File: deepchem/models/torch_models/gnn.py (hunk 5 above)
Comment (tonydavis629) [reply deepchem/models/torch_models/gnn.py:838]: done

File: deepchem/models/losses.py (hunk 6 above)
Comment (tonydavis629) [reply deepchem/models/losses.py:1115]: done

File: deepchem/models/torch_models/gnn.py (hunk 7 above)
Comment (tonydavis629) [reply deepchem/models/torch_models/gnn.py]: done

File: deepchem/models/torch_models/gnn.py (hunk 8 above)
Comment (tonydavis629) [reply deepchem/models/torch_models/gnn.py:184]: done. I think we should maybe have a section in `layers.rst` for graph layers, or maybe torch geometric layers.

**Review by rbharath — APPROVED** [review]
LGTM. Feel free to merge in once CI looks clear
